### PR TITLE
RC75: avoid ambiguous RTT calcs and enable udt fast retransmit

### PR DESCRIPTION
**Pull request by birarda**
https://highfidelity.manuscript.com/f/cases/19289/Some-users-see-persistent-moving-white-spheres-for-other-avatars

### CODE REVIEW

**Comment by birarda** [comment]
#### Test Plan

Test in engine-dev with 300 bots and 10% packet loss between your Interface client and the avatar-mixer (using clumsy). With the master avatar mixer, you should see white spheres or t-posed woody avatars for a long period of time when the bots are restarted. You may also see a falsely inflated count of avatars from lost kill packets. With the avatar mixer built from this PR, you should see the white spheres and t-posted woody avatars only for a short period of time. You should also see the number of avatars settle relatively quickly to the right amount.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14515/HighFidelity-Beta-PR14304-76ff2ac.apk

File: libraries/networking/src/udt/TCPVegasCC.h [hunk 1]
```diff
@@ -30,18 +30,31 @@ class TCPVegasCC : public CongestionControl {
...
+        SentPacketData(SequenceNumber seqNum, p_high_resolution_clock::time_point tPoint)
+            : sequenceNumber(seqNum), timePoint(tPoint) {};
+        
+        SequenceNumber sequenceNumber;
+        p_high_resolution_clock::time_point timePoint;
+        bool wasResent { false };
+    };
+
+    using PacketTimeList = std::vector<SentPacketData>;
+    PacketTimeList _sentPacketDatas; // association of sequence numbers to sent time, for RTT calc
```
Comment (Atlante45) [inline libraries/networking/src/udt/TCPVegasCC.h]: This should always be contiguous so you could have something like this:
```cpp
class PacketTimeList {
    struct SentPacketData {
        p_high_resolution_clock::time_point timePoint;
        bool wasResent { false };
    };
    using PacketTimeList = std::vector<SentPacketData>;

    SequenceNumber _firstSequenceNumber;
    PacketTimeList _sentPacketDatas; 
};
```

File: libraries/networking/src/udt/TCPVegasCC.cpp [hunk 2]
```diff
@@ -27,117 +27,155 @@ TCPVegasCC::TCPVegasCC() {
...
-        ++_numACKs;
+    if (!wasDuplicateACK && it != _sentPacketDatas.end()) {
+        // check if we can unambigiously calculate an RTT from this ACK
 
-        // keep track of the lowest RTT during connection
-        _baseRTT = std::min(_baseRTT, lastRTT);
+        // for that to be the case,
+        // any of the packets this ACK covers (from the current ACK back to our previous ACK)
+        // must not have been re-sent
+        auto ambigiousMatch = std::find_if(_sentPacketDatas.begin(), _sentPacketDatas.end(),
```
[the remark this answers is not in the document]
Comment (birarda) [reply libraries/networking/src/udt/TCPVegasCC.cpp]: Will change

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14519/HighFidelity-Beta-PR14304-5fbb729.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14519/HighFidelity-Beta-PR14304-5fbb729.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14519/HighFidelity-Beta-Interface-PR14304-5fbb729.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14519/HighFidelity-Beta-PR14304-5fbb729.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14519/HighFidelity-Beta-Interface-PR14304-5fbb729.exe

**Comment by antoninahifi** [comment]
currently testing

**Comment by antoninahifi** [comment]
@conklin94122 PASS

**Pull request merged**